pilot/cmd/istioctl/convert/destination.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2017 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package convert
+
+import (
+	routing "istio.io/api/routing/v1alpha1"
+	routingv2 "istio.io/api/routing/v1alpha2"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+)
+
+func convertDestinationPolicies(configs []model.Config) []model.Config {
+	policies := make([]*routing.DestinationPolicy, 0)
+	for _, config := range configs {
+		if config.Type == model.DestinationPolicy.Type {
+			policies = append(policies, config.Spec.(*routing.DestinationPolicy))
+		}
+	}
+
+	destinationRuleByHost := make(map[string]*routingv2.DestinationRule)
+
+	for _, policy := range policies {
+		host := convertIstioService(policy.Destination)
+		if destinationRuleByHost[host] == nil {
+			destinationRuleByHost[host] = &routingv2.DestinationRule{Name: host}
+		}
+		destinationRule := destinationRuleByHost[host]
+		destinationRule.Subsets = append(destinationRule.Subsets, convertDestinationPolicy(policy))
+	}
+
+	// rewrite traffic policies where len(subsets) == 1
+	for _, rule := range destinationRuleByHost {
+		if len(rule.Subsets) == 1 {
+			rule.TrafficPolicy = rule.Subsets[0].TrafficPolicy
+			rule.Subsets = nil
+		}
+	}
+
+	out := make([]model.Config, 0)
+	for _, rule := range destinationRuleByHost {
+		// TODO: ConfigMeta needs to be aggregated from source configs somehow
+		out = append(out, model.Config{
+			ConfigMeta: model.ConfigMeta{
+				Type:            model.V1alpha2RouteRule.Type,
+				Name:            "FIXME",
+				Namespace:       "FIXME",
+				Domain:          "FIXME",
```

Comment:
I think to be correct we'd have to create a route rule per `(host, namespace, domain)`, instead of just `host`. However, given the nature of the tool, I think it'd also be fair to just pick the namespace + domain of one of the source `DestinationPolicies` (e.g. the first we touch for that host) and print a warning to STDERR saying it merged rules across namespaces and advising to only process files with objects belonging to a single namespace at a time.

Reply:
That's a good suggestion, thanks